Rework suggested imports functionality

andysellick opened the pull request:
## What
Modifies the logic behind the suggested imports functionality of the component guide to be more specific when dealing with includes that look like components, but aren't.

## Why
This is causing a 500 error in content-publisher, see https://github.com/alphagov/govuk_publishing_components/issues/1570

## Visual Changes
None.

andysellick commented:
@alex-ju @kevindew I don't have a locally running version of content-publisher, so I haven't been able to test this thoroughly yet, but will keep trying. The change to the dummy app should mean the tests give some confidence this change has worked, but I want to be sure.

alex-ju commented:
> @alex-ju @kevindew I don't have a locally running version of content-publisher, so I haven't been able to test this thoroughly yet, but will keep trying. The change to the dummy app should mean the tests give some confidence this change has worked, but I want to be sure.

I can confirm the 500 error is gone with this change. Will try and make time to review the code too.

andysellick commented:
@alex-ju which pages? You mean the detail of the component pages within the component guide? Is that only in content-publisher? I'm online now so able to check this properly.

andysellick commented on app/controllers/govuk_publishing_components/component_guide_controller.rb:
```diff
@@ -86,24 +86,30 @@ def components_in_use
       files.each do |file|
         data = File.read(file)
-        matches << data.scan(/(govuk_publishing_components\/components\/[a-z_-]+)/)
+        matches << data.scan(/(govuk_publishing_components\/components\/[\/a-zA-Z_-]+['"]{1})/)
```
Thanks. Actually I've realised I don't need that bit on the end of the regexp at all, so I've removed both bits 👍

kevindew replied:
Cool, is that because of the check for the file existing? I guess it now expects something like thumbnail.svg to be a component but just doesn't find a thumbnail.html.erb?

Probably still worth using the look behind to avoid you needing the `m.gsub("govuk_publishing_components/components/", "")` twice in the code.

andysellick replied:
Ah, of course. Your regex-fu is stronger than mine. Thanks, that's much better.

andysellick commented on app/controllers/govuk_publishing_components/component_guide_controller.rb:
```diff
@@ -132,9 +137,12 @@ def component_has_js_file(component)
       data = File.read(filename)
-      match = data.scan(/(govuk_publishing_components\/components\/[\/a-z_-]+)/)
-      match.flatten.uniq.map(&:to_s).sort.map { |m| m.gsub("govuk_publishing_components/components/", "") }
+      match = data.scan(/(govuk_publishing_components\/components\/[\/a-zA-Z_-]+['"]{1})/)
```
Good point, done.

kevindew replied:
Thanks, sorry to be a pain but it seems somewhat unconventional as an instance variable and would seem more conventional using a constant as per: https://github.com/alphagov/govuk_publishing_components/blob/aa86284e7081788b6d22c50cfb70ef3992aa7dca/lib/govuk_publishing_components/presenters/machine_readable/potential_search_action_schema.rb#L8

andysellick replied:
Fair enough, have changed.

andysellick commented:
@kevindew thanks for the review, think I've addressed your comments although I still haven't got content-publisher running locally to test the bug @alex-ju found on Friday. 

Yes, this feature is mainly to aid adoption of the individual component model. After that's complete it might still be useful for ongoing maintenance, so we might remove it, or we might just put it on another page in the component guide so it doesn't slow down the homepage so much. Good to know the numbers, thanks for that.
